### PR TITLE
Add a bunch of Improvements (a bit ad-hoc)

### PR DESCRIPTION
**Pull request by aldelaro5**
For the past week, I found that I could actually get very close to recovering all the shaders in Bug Fables: The Everlasting Sapling thanks to the yaml assets export option in AssetRipper. USC was more advanced than the AssetRipper decompiler, so I tried to fix things and implement things on usc enough to get to a point where all shaders are recoverded.

I succeeded, but it took a lot of changes. I tried my best to segment the commits with high level explanations, but unfortunately, too much happened for this to be the cleanest. I am especially annoyed at ShaderProcessor being hard to follow from the diff, but it will do for now.

The commits list gives a good high level overview of what I had to do, but the biggest one is worth talking about: shaders variants.

The biggest issue I was running into was that either decompilers was not including all the variants needed. Some of these are needed for the materials to receive lighting or shadows so it made a very big difference to have them. I solved this by completely changing the structure of how we decompile sub programs.

First, we no longer try to pair them, we just have one basket with everything in it. This means it will decompile all vertex followed by all fragments. I also had to make each shader decompile their own struct, cbuffers and textures because I found they can differ between variants. Unfortunately, this means this is EXTREMELY inneficient: files grows by thousands of lines in worst cases because a lot of the code or declarations gets repeated. Further optimisations can be done later to merge the common stuff, but I just cared for accuracy at the moment.

Second, keywords. How this is done is we first build a list of multi_compile and shader_feature directives to map out the list of keywords combinations we want and select the variant via an `#if`, `#elif`, `#else` `#endif` chain. Due to how this chain works logically, it is best to have every shaders sorted from most specific to least (after sorting them by types) so it allows us to test just the keywords that are enabled in the ifs and add a falback else clause on the last one so we guarantee one shader is selected.

Unfortunately, this keywords mapping isn't 1:1: Unity doesn't allow to have a specific set of keywords to be a variant, it only allows to declare sets where each will be one. This complicates our lives because we now have to specify enough directive to include everything, but some overgenerations is unavoidable. The sweetspot I got was the following:

1. If a keyword is always there, put a multi_compile with just that keyword which forces it on
2. Excluding the mandatory keywords found above, considering all the combinations with the least amounts of keywords, make multi_compile directives for each indexes until we cover the least amount of keywords possible
3. For everything else, put a shader_feature with that keyword which makes it completely optional

With all the changes done, I was able to fully recover all shaders from the game.

### CODE REVIEW

File: USCSandbox/Program.cs [hunk 1]
```diff
@@ -34,16 +54,39 @@ static void Main(string[] args)
...
+                    continue;
+                
+                var shaderBf = manager.GetExtAsset(afileInst, 0, shader.PathId).baseField;
+                if (shaderBf == null)
+                {
+                    Console.WriteLine("Shader asset not found.");
+                    return;
+                }
+                var shaderProcessor = new ShaderProcessor(shaderBf, ver, shaderPlatform);
+                bool fileNameExists = files.TryGetValue(shader.PathId, out string? name);
```
Comment (nesrak1) [inline USCSandbox/Program.cs:69]: You don't need the container string for the shader name. Just used m_ParsedForm:
```suggestion
                var name = shaderBf["m_ParsedForm"]["m_Name"].AsString;
```

File: USCSandbox/UltraShaderConverter/USIL/Metadders/USILCBufferMetadder.cs [hunk 2]
```diff
@@ -77,6 +79,13 @@ private static void UseMetadata(USILOperand operand, ShaderParams shaderParams)
                             {
                                 maskIndex %= 4;
                             }
+
+                            if (param.ArraySize > 1)
```
Comment (nesrak1) [inline USCSandbox/UltraShaderConverter/USIL/Metadders/USILCBufferMetadder.cs]: Matrices have different behavior.
```suggestion
                            else if (param.ArraySize > 1)
```

Below, you can replace the cbParams.Count == 1 code with this to handle matrix arrays:
```cs
else if (cbParams.Count == 1)
{
    ConstantBufferParameter param = cbParams.First();

    operand.arrayIndex -= param.Index / 16;

    // apparently switch has column long, whereas directx has row long
    int maxRowOrColumnLength = Math.Max(param.Rows, param.Columns);

    if (param.IsMatrix)
    {
        if (param.ArraySize > 0)
        {
            // example of 4x4 matrix: matrix[5] -> matrix[1][1] (matrix[1]._m01._m11._m21._m31)
            operand.arrayRelative = new USILOperand(operand.arrayIndex / maxRowOrColumnLength);
            operand.arrayIndex %= maxRowOrColumnLength;
        }

        operand.operandType = USILOperandType.Matrix;
        operand.transposeMatrix = true;
    }

    operand.mask = cbMasks.ToArray();
    operand.metadataName = param.ParamName;
    operand.metadataNameAssigned = true;
    operand.metadataNameWithArray = param.ArraySize > 1;

    if (cbMasks.Count == maxRowOrColumnLength && !param.IsMatrix)
    {
        operand.displayMask = false;
    }
}
```

File: USCSandbox/UltraShaderConverter/USIL/Metadders/USILSamplerMetadder.cs [hunk 3]
```diff
@@ -19,7 +19,7 @@ public bool Run(UShaderProgram shader, ShaderParams shaderParams)
                         //    p => p.SamplerIndex == operand.registerIndex
                         //);
                         TextureParameter? texParam = shaderParams.TextureParameters.FirstOrDefault(
-                            p => p.Index == operand.registerIndex
+                            p => p.SamplerIndex == operand.registerIndex
                         );
 
                         if (texParam == null)
```
Comment (nesrak1) [inline USCSandbox/UltraShaderConverter/USIL/Metadders/USILSamplerMetadder.cs:22]: Even with SamplerIndex this still seems to fail sometimes, but in those cases it seems we have a -1 SamplerIndex. Let's add that just in case:
```suggestion
                        if (texParam == null)
                        {
                            // fallback to -1 if it exists
                            texParam = shaderParams.TextureParameters.FirstOrDefault(
                                p => p.SamplerIndex == -1
                            );

                            if (texParam == null)
```

File: USCSandbox/Processor/ShaderProcessor.cs [hunk 4]
```diff
@@ -210,12 +420,12 @@ public string Process()
                 bool nonGlobalCbuffer = cbuffer.Name != "$Globals";
                 int cbufferIndex = shaderParams.ConstantBuffers.IndexOf(cbuffer);
 
-                if (nonGlobalCbuffer)
-                {
-                    sb.Append(new string(' ', depth * 4)); // todo: new stringbuilder
-                    sb.AppendLine($"CBUFFER_START({cbuffer.Name}) // {cbufferIndex}");
-                    depth++;
-                }
+                // if (nonGlobalCbuffer)
```
Comment (nesrak1) [inline USCSandbox/Processor/ShaderProcessor.cs]: This is important to keep since users can also write CBUFFER_START. We can't just comment all of these out. There's two things I noticed:
1. Sometimes the name already exists. This just means we need to update the list we have of things to skip.
2. Sometimes the cbuffer is empty. We should just do something like this:
```cs
if (!wroteCbufferHeaderYet && nonGlobalCbuffer)
{
    sb.Append(new string(' ', depth * 4)); // todo: new stringbuilder
    sb.AppendLine($"CBUFFER_START({cbuffer.Name}) // {cbufferIndex}");
    depth++;
    wroteCbufferHeaderYet = true;
}
```

File: USCSandbox/Processor/ShaderProcessor.cs [hunk 5]
```diff
@@ -235,23 +445,27 @@ public string Process()
...
                     }
                 }
 
-                if (nonGlobalCbuffer)
-                {
-                    depth--;
-                    sb.Append(new string(' ', depth * 4));
-                    sb.AppendLine("CBUFFER_END");
-                }
+                // if (nonGlobalCbuffer)
```
Comment (nesrak1) [inline USCSandbox/Processor/ShaderProcessor.cs]: ditto

File: USCSandbox/UltraShaderConverter/UShader/DirectX/DirectXProgramToUSIL.cs [hunk 6]
```diff
@@ -1288,8 +1316,8 @@ private void HandleDp2(SHDRInstruction inst)
 
             int[] mask = new int[] { 0, 1 };
             FillUSILOperand(dest, usilDest, dest.swizzle, false);
-            FillUSILOperand(src0, usilSrc0, MapMask(mask, src0.swizzle), false);
-            FillUSILOperand(src1, usilSrc1, MapMask(mask, src1.swizzle), false);
+            FillUSILOperand(src0, usilSrc0, src0.swizzle.Length >= 2 ? src0.swizzle[..2] : src0.swizzle, false);
```
Comment (nesrak1) [inline USCSandbox/UltraShaderConverter/UShader/DirectX/DirectXProgramToUSIL.cs]: I don't see how this is any different than what MapMask does. Also, is there any case where the swizzle would be under 2 for a dp2? I would rather use MapMask for consistency.

**Review by nesrak1 — CHANGES_REQUESTED** [review]
If/when you go to clean this up, go ahead and remove all the code comments (don't remove the BUILTIN_CBUFFER_NAMES comment though). It's git, so we can always look back at them.

File: USCSandbox/Program.cs (hunk 1 above)
Comment (aldelaro5) [reply USCSandbox/Program.cs:69]: This is incorrect: I want I to replicate what assetripper did here because it allows me to very quickly copy paste the file into the unity project and test them. This is also not practical to do because these are how the shaders are named in editor, but NOT how it is placed in the project's folder which is what I care about here.

Comment (nesrak1) [reply USCSandbox/Program.cs:69]: I would at least suggest it as a fallback. Most of the shaders I tested didn't have a container causing me to have hundreds of "unnamed shader" files

Comment (aldelaro5) [reply USCSandbox/Program.cs:69]: The unanmed are actually unity's integraded shaders

They don't matter because a project never has them, it just comes with the editor and is compiled in the assets

Comment (aldelaro5) [reply USCSandbox/Program.cs:69]: Ended up adding a falback for those

File: USCSandbox/UltraShaderConverter/UShader/DirectX/DirectXProgramToUSIL.cs (hunk 6 above)
Comment (aldelaro5) [reply USCSandbox/UltraShaderConverter/UShader/DirectX/DirectXProgramToUSIL.cs]: Tested, found it only did one good change to the decomp of one shader so this must have been done for precaution, but it doesn't appear this was needed so I reverted it

**Comment by aldelaro5** [comment]
@nesrak1 done, also removed all code comments
